src/simulator.cpp

```diff
@@ -33,118 +38,140 @@ void Simulator::RunSimulation() {
 }
 
 bool Simulator::_CacheHandler(inst_t inst) {
-    bool is_load(false), is_store(false), is_space(false);
+    addr_t next_addr = Cvt2AddrBits(inst.addr_raw);
+
     // Determine what kind of the instruction
     switch (inst.op) {
     case I_LOAD:
-        is_load = true;
+        _Load(next_addr);
         break;
     case I_STORE:
-        is_store = true;
+        _Store(next_addr);
         break;
     case I_NONE:
-        is_space = true;
+        ++_counter.space;
         break;
+    default:
+        std::cerr << "Unexpected error in _CacheHandler()" << std::endl;
+        std::cerr << "ERROR line: " << inst.addr_raw << std::endl;
+        return false;
     }
+    return true;
+}
 
-    addr_t next_addr = Cvt2AddrBits(inst.addr_raw);
-
-    this->main_cache->Ready(next_addr);
-
-    // Handle the address
-    if (is_load) {
-        ++_counter.access;
-        ++_counter.load;
-
-        if (this->main_cache->IsHit()) {
+void Simulator::_Load(const addr_t &addr) {
+    bool _is_hit(false);
+    ++_counter.access;
+    ++_counter.load;
+
+    if (_multi_level) {
+        for (auto _cache = _cache_hierarchy_list.begin();
+             _cache != _cache_hierarchy_list.end() && !_is_hit; ++_cache) {
+            if (_cache->Get(addr)) {
+                _is_hit = true;
+                ++_counter.load_hit;
+                ++_counter.hit_in_main;
+            } else {
+                _is_hit = false;
+                _cache->Set(addr);
+            }
+        }
+    } else {
+        if (_cache_hierarchy_list[0].Get(addr)) {
+            _is_hit = true;
             ++_counter.load_hit;
             ++_counter.hit_in_main;
         } else {
-            this->main_cache->_Read(next_addr);
+            _is_hit = false;
+            _cache_hierarchy_list[0].Set(addr);
         }
-    } else if (is_store) {
-        ++_counter.access;
-        ++_counter.store;
+    }
+}
 
-        if (this->main_cache->_IsHit(next_addr)) {
+void Simulator::_Store(const addr_t &addr) {
+    bool _is_hit(false);
+    ++_counter.access;
+    ++_counter.store;
+
+    if (_multi_level) {
+        for (auto _cache = _cache_hierarchy_list.begin();
+             _cache != _cache_hierarchy_list.end() && !_is_hit; ++_cache) {
+            if (_cache->Get(addr)) {
+                _is_hit = true;
+                ++_counter.store_hit;
+                ++_counter.hit_in_main;
+            } else {
+                _is_hit = false;
+                _cache->Set(addr);
+            }
+        }
+    } else {
+        if (_cache_hierarchy_list[0].Get(addr)) {
+            _is_hit = true;
             ++_counter.store_hit;
             ++_counter.hit_in_main;
-
         } else {
-            this->main_cache->_Read(next_addr);
+            _is_hit = false;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable '_is_hit' is assigned a value that is never used.](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708)